Cython/Compiler/Code.py

```diff
@@ -2129,7 +2129,7 @@ def put_xdecref_clear(self, cname, type, nanny=True, clear_before_decref=False):
                          clear=True, clear_before_decref=clear_before_decref)
 
     def _put_decref(self, cname, type, nanny=True, null_check=False,
-                    have_gil=True, clear=False, clear_before_decref=False):
+                    have_gil='gil', clear=False, clear_before_decref=False):
```

Comment:
I'm not hugely keen on this - it seems to be complicating what should be a simple flag, and (unless I've missed something) this seems like you could just use "False" for the on-device bits.

Reply:
Unfortunately not. Even with nogil Cython generates checks that are useful on a CPU but not so much (or better impractical) on a GPU/device. See below diff (put_incref_memoryviewslice).

Reply:
Ah - fair enough. I can see why you need it. Two possible options to consider:

1. Could you change the utility code to make the `__PYX_INC_MEMVIEW` macro evaluate to nothing on the device?
2. I have [a PR possibly pending](https://github.com/cython/cython/pull/3377) that tries to re-arrange how the various put_xxxref functions are generated. I'm not sure you'd be well-advised to rely on a PR that may not be accepted or may have to be modified to be accepted. However, the advantage would be that you could make a dedicated `DeviceMemviewSlice` type that inherits from the normal `MemviewSlice` type but doesn't generate reference counting. (If you were to consider this option I'd probably leave your code as is for now, and see what happens with my PR)

There may of course be other, better options which aren't obvious to me.

Reply:
@scoder I don't think I know how this would look like. Could you give a an idea how this would work?